controller.py

```diff
@@ -81,46 +83,157 @@
-        # (i.e. URN lookup) return None instead of an error.
-        return None
-    return INVALID_CREDENTIALS
+    A simple grouping of controllers.
+    """
 
+    def __init__(self,_db):
+        self._db = _db
+        self.heartbeat = HeartbeatController()
+        self.canonicalization =  CanonicalizationController(self._db)
+        self.index = IndexController(self._db)
+        self.urn_lookup = URNLookupController(self._db)
+        self.catalog = CatalogController(self._db)
+        self.integration = IntegrationClientController(self._db)
 
-def collection_from_details(_db, client, collection_details):
-    if not (client and isinstance(client, IntegrationClient)):
-        return None
+    @classmethod
+    def authenticated_client_from_request(cls, _db, required=True):
+        """Look up the IntegrationClient that's making the active
+        request.
 
-    if collection_details:
-        # A DataSource may be sent for collections with the
-        # ExternalIntegration.OPDS_IMPORT protocol.
-        data_source_name = request.args.get('data_source')
-        if data_source_name:
-            data_source_name = urllib.unquote(data_source_name)
+        This function is used by controller code and by a decorator
+        applied to Flask routes.
 
-        collection, ignore = Collection.from_metadata_identifier(
-            _db, collection_details, data_source=data_source_name
-        )
-        return collection
-    return None
+        :param required: True if a lack of authentication or bad
+                         authentication is an error condition.
 
+        :return: An IntegrationClient if one could be authenticated;
+                 otherwise None (if authentication is not required)
+                 or a ProblemDetail.
+        """
+        # Start with the assumption that there is no authenticated
+        # IntegrationClient.
+        flask.request.authenticated_client = None
+
+        header = flask.request.headers.get('Authorization')
+        if header:
+            if not 'bearer' in header.lower():
+                # No authentication mechanism other than a bearer token
+                # is supported
+                return INVALID_CREDENTIALS
+
+            # We can hopefully use the provided bearer token to
+            # authenticate an IntegrationClient.
+            try:
+                shared_secret = base64.b64decode(header.split(' ')[1])
+            except TypeError, e:
+                # The bearer token is ill-formed.
+                return INVALID_CREDENTIALS
+            client = IntegrationClient.authenticate(_db, shared_secret)
+
+            if client:
+                # Success!
+                flask.request.authenticated_client = client
+                return client
+            else:
+                # The bearer token doesn't identify any particular
+                # IntegrationClient.
+                return INVALID_CREDENTIALS
```

Comment:
Previously, if you passed in bad credentials to do something that allowed but didn't require authentication, you would perform the action as though you had not authenticated. I think it's better if your bad credentials are rejected; that way you know there's a problem.